src/flat_list.jsx

```diff
@@ -15,57 +15,116 @@ class FlatList extends React.Component {
-        clearTimeout(this.timeoutId);
-        const hasMore =
-          items.length + batchCount <= data.length ||
-          (data.length - items.length <= batchCount && data.length !== items.length);
-        if (hasMore) {
-          this.timeoutId = setTimeout(() => {
-            this.setState(prevState => ({
-              items: [...items, ...data.slice(prevState.items.length, prevState.items.length + batchCount)]
-            }));
-          }, timeoutMilliSec);
-        }
+        this.loadMoreItems();
+      } else {
+        this.setState({ items: [] });
+      }
+    } else if (isScroll) {
+      if (oldData.length == data.length && itemChecker({ data: oldData }, { data })) {
+        if (items.length == 0)
+          this.setState({
+            items: data.slice(0, batchCount)
+          });
       } else {
         this.setState({ items: [] });
       }
     }
+    if (itemToScrollPosition >= 0 && itemToScrollPosition != null) {
+      if (itemToScrollPosition > items.length) {
+        this.loadMoreItems();
+      } else {
+        this.scrollToElement(itemToScrollPosition);
+      }
+    }
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener("scroll", this.handleOnScroll);
+  }
+
+  loadMoreItems() {
+    const { data, batchCount } = this.props;
+    const { items } = this.state;
+    clearTimeout(this.timeoutId);
+    const hasMore =
+      items.length + batchCount <= data.length ||
+      (data.length - items.length <= batchCount && data.length !== items.length);
+    if (hasMore) {
+      this.timeoutId = setTimeout(() => {
+        this.setState(prevState => ({
+          items: [...items, ...data.slice(prevState.items.length, prevState.items.length + batchCount)]
+        }));
+      }, timeoutMilliSec);
+    }
+  }
+
+  scrollToElement(itemPosition) {
+    const { onScrollEnd } = this.props;
+    const { items } = this.state;
+    const referenceElement = itemPosition == items.length - 1 ? this.lastItemRef.current : this.refs[`Ref-${itemPosition}`];
+    clearTimeout(this.scrollEnd);
+    referenceElement.scrollIntoView({
+      behavior: "smooth",
+      block: "center"
+    });
+    this.scrollEnd = setTimeout(() => onScrollEnd(), 3000);
   }
 
   render() {
-    const { data, renderItem, itemKey, isVirtual } = this.props;
+    const { data, renderItem, itemKey, isVirtual, isScroll, ListItemType } = this.props;
     const { items } = this.state;
-    return (isVirtual ? items : data).map(item => (
-      <ItemRenderer
-        key={`List-${typeof item == "object" ? item[itemKey] : item}`}
-        item={item}
-        renderItem={renderItem}
-      />
+    const lastIndex = items.length - 1;
+    return (isVirtual || isScroll ? items : data).map((item, idx) => (
+      <React.Fragment key={`List-${typeof item == "object" ? item[itemKey] : item}`}>
+        <ItemRenderer item={item} renderItem={renderItem} />
+        <ListItemType ref={lastIndex == idx ? this.lastItemRef : `Ref-${idx}`}></ListItemType>
```

Comment:
Why Do We need a condition Now for last item in list ? All the list items will have this ref.
Please Use Better ref prefix. (`Ref-` => `separator-ref`) ..... and html attr & values are all small case.

Reply:
Depending on index is quite fishy in react